diff -urN sys/dev/virtio.ori/block/virtio_blk.h sys/dev/virtio/block/virtio_blk.h
--- sys/dev/virtio.ori/block/virtio_blk.h	2020-03-19 20:20:23.688780000 -0700
+++ sys/dev/virtio/block/virtio_blk.h	2020-03-19 23:17:51.681970000 -0700
@@ -1,4 +1,6 @@
 /*-
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
  * This header is BSD licensed so anyone can use the definitions to implement
  * compatible drivers/servers.
  *
@@ -32,17 +34,22 @@
 #define _VIRTIO_BLK_H
 
 /* Feature bits */
-#define VIRTIO_BLK_F_BARRIER	0x0001	/* Does host support barriers? */
 #define VIRTIO_BLK_F_SIZE_MAX	0x0002	/* Indicates maximum segment size */
 #define VIRTIO_BLK_F_SEG_MAX	0x0004	/* Indicates maximum # of segments */
 #define VIRTIO_BLK_F_GEOMETRY	0x0010	/* Legacy geometry available  */
 #define VIRTIO_BLK_F_RO		0x0020	/* Disk is read-only */
 #define VIRTIO_BLK_F_BLK_SIZE	0x0040	/* Block size of disk is available*/
-#define VIRTIO_BLK_F_SCSI	0x0080	/* Supports scsi command passthru */
-#define VIRTIO_BLK_F_WCE	0x0200	/* Writeback mode enabled after reset */
+#define VIRTIO_BLK_F_FLUSH	0x0200	/* Flush command supported */
 #define VIRTIO_BLK_F_TOPOLOGY	0x0400	/* Topology information is available */
 #define VIRTIO_BLK_F_CONFIG_WCE 0x0800	/* Writeback mode available in config */
+#define VIRTIO_BLK_F_MQ 	0x1000 	/* Support more than one vq */
 
+/* Legacy feature bits */
+#define VIRTIO_BLK_F_BARRIER	0x0001	/* Does host support barriers? */
+#define VIRTIO_BLK_F_SCSI	0x0080	/* Supports scsi command passthru */
+
+/* Old (deprecated) name for VIRTIO_BLK_F_FLUSH. */
+#define VIRTIO_BLK_F_WCE VIRTIO_BLK_F_FLUSH
 #define VIRTIO_BLK_ID_BYTES	20	/* ID string length */
 
 struct virtio_blk_config {
@@ -64,15 +71,23 @@
 
 	/* Topology of the device (if VIRTIO_BLK_F_TOPOLOGY) */
 	struct virtio_blk_topology {
+		/* exponent for physical block per logical block. */
 		uint8_t physical_block_exp;
+		/* alignment offset in logical blocks. */
 		uint8_t alignment_offset;
+		/* minimum I/O size without performance penalty in logical
+		 * blocks. */
 		uint16_t min_io_size;
+		/* optimal sustained I/O size in logical blocks. */
 		uint32_t opt_io_size;
 	} topology;
 
 	/* Writeback mode (if VIRTIO_BLK_F_CONFIG_WCE) */
-	uint8_t writeback;
+	uint8_t wce;
+	uint8_t unused;
 
+	/* Number of vqs, only available when VIRTIO_BLK_F_MQ is set */
+	uint16_t num_queues;
 } __packed;
 
 /*
@@ -105,7 +120,11 @@
 /* ID string length */
 #define VIRTIO_BLK_ID_BYTES	20
 
-/* This is the first element of the read scatter-gather list. */
+/*
+ * This comes first in the read scatter-gather list.
+ * For legacy virtio, if VIRTIO_F_ANY_LAYOUT is not negotiated,
+ * this is the first element of the read scatter-gather list.
+ */
 struct virtio_blk_outhdr {
 	/* VIRTIO_BLK_T* */
 	uint32_t type;